File: kth_smallest_element.java
// Time Complexity : O(n)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No


// Your code here along with comments explaining your approach
class Solution {
    public int kthSmallest(int[][] matrix, int k) {
        int n = matrix.length;
        int low = matrix[0][0];
        int high = matrix[n - 1][n - 1];
        while (low < high) {
            int mid = low + (high - low) / 2;
            int count = getCount(matrix, mid);
            if (count < k) {
                low = mid + 1;
            } else {
                high = mid;
            }
        }
        return low;
    }

    public int getCount(int[][] matrix, int mid) {
        int count = 0;
        int j = matrix.length - 1;
        for (int i = 0; i < matrix.length; i++) {
            while (j >= 0 && matrix[i][j] > mid)
                j--;
            count += j + 1;
        }
        return count;
    }
} 

File: kthsmallestelement.py
# Time Complexity :O(NM log k)
# Space Complexity : O(k)
# Did this code successfully run on Leetcode :Yes
from heapq import heappush as insert
from heapq import heappop as remove

class Solution:
    def kthSmallest(self, matrix: List[List[int]], k: int) -> int:
        self.heap = []
        curr_size = 0
        for i in range(0,len(matrix)):
            for j in range(0,len(matrix[0])):
                insert(self.heap, -1 * matrix[i][j])
                curr_size += 1
                if curr_size > k :
                    remove(self.heap)
        return -1 * remove(self.heap)